Implement Checkpoint, Restart, and Plotfile for pure SOA

atmyers opened the pull request:
Follow-up to #2878.

The proposed changes:
- [x] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

atmyers committed:
[WIP] Implement Checkpoint, Restart, and Plotfile for pure SOA

atmyers committed:
add test

atmyers committed:
bug, replace AMREX_SPACEDIM by 2

atmyers committed:
Merge branch 'development' into CheckpointRestartPureSOA

ax3l commented on Tests/Particles/CheckpointRestartSOA/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+   return()
+endif ()
+
+set(_sources     main.cpp)
+set(_input_files inputs  )
+
+setup_test(_sources _input_files)
+
+unset(_sources)
+unset(_input_files)
```
```suggestion
foreach(D IN LISTS AMReX_SPACEDIM)
    set(_sources     main.cpp)
    set(_input_files inputs  )

    setup_test(${D} _sources _input_files)

    unset(_sources)
    unset(_input_files)
endforeach()
```

atmyers committed:
Update Tests/Particles/CheckpointRestartSOA/CMakeLists.txt

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented on Tests/Particles/CheckpointRestartSOA/main.cpp:
```diff
@@ -0,0 +1,221 @@
+
+    Vector<DistributionMapping> dmap(nlevs);
+
+    Vector<std::unique_ptr<MultiFab> > mf(nlevs);
```
```suggestion
    // write some mesh data too, because tools like yt expect them to be there
    Vector<std::unique_ptr<MultiFab> > mf(nlevs);
```

ax3l commented on Tests/Particles/CheckpointRestartSOA/main.cpp:
```diff
@@ -0,0 +1,221 @@
+        {
+            amrex::Print() << "working on comp " << icomp << "\n";
+            auto sm_new = amrex::ReduceSum(newPC,
+                [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+                {
+                    return p.rdata(1);
+                });
+
+            auto sm_old = amrex::ReduceSum(myPC,
+                [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+                {
+                    return p.rdata(1);
```
Let us varify id/cpu and position data, too.

ax3l committed:
CP/Restart: Keep Struct on Disk

For backwards compatibility with tools like yt and VisIt,
we keep the id, cpu and positions in a struct on disk for
plotfiles.

atmyers committed:
Fix async out for pure SOA and add test

atmyers committed:
merging

atmyers committed:
Test all comps

atmyers committed:
Merge branch 'CheckpointRestartPureSOA' of github.com:atmyers/amrex into CheckpointRestartPureSOA

ax3l committed:
Add Test MF Comment

ax3l committed:
Cleanup: Variable Shadowing

ax3l committed:
Async IO Test Separately

ax3l reviewed (approved):
:rocket: :sparkles:

atmyers committed:
ignore unused

atmyers committed:
ignored->ignore
